ffi/derive/src/arg.rs

```diff
@@ -0,0 +1,494 @@
+//! Logic related to FFI function argument. Visitor implementation visits the given type
+//! and collects the FFI conversion information into the [`Arg`] struct
+#![allow(clippy::unimplemented)]
+
+use proc_macro2::Span;
+use proc_macro_error::{abort, OptionExt};
+use quote::quote;
+use syn::{
+    parse_quote, visit::Visit, visit_mut::VisitMut, Ident, Path, PathArguments::AngleBracketed,
+    Type,
+};
+
+use crate::get_ident;
+
+/// Struct representing a method/function argument
+#[derive(Clone)]
+pub struct Arg {
+    /// Name of the function/method argument
+    pub name: Ident,
+
+    /// Rust type
+    pub src_type: Type,
+    /// FFI compliant type
+    pub ffi_type: Type,
+
+    /// Conversion statement from Rust to FFI compliant type
+    pub src_to_ffi: syn::Stmt,
+    /// Conversion statement from FFI compliant to Rust type
+    pub ffi_to_src: syn::Stmt,
+}
+
+impl Arg {
+    pub fn handle(self_ty: &Path, src_type: Type) -> Self {
+        let handle_name = Ident::new("__handle", Span::call_site());
+        Self::new(self_ty, handle_name, src_type, true)
+    }
+
+    pub fn input(self_ty: &Path, name: Ident, src_type: Type) -> Self {
+        Self::new(self_ty, name, src_type, true)
+    }
+
+    pub fn output(self_ty: &Path, name: Ident, src_type: Type) -> Self {
+        Self::new(self_ty, name, src_type, false)
+    }
+
+    fn new(self_ty: &Path, name: Ident, src_type: Type, is_input: bool) -> Self {
+        let mut visitor: TypeVisitor = TypeVisitor::new(self_ty, &name, is_input);
+        visitor.visit_type(&src_type);
+        let mut ffi_type = visitor.ffi_type.expect_or_abort("Defined");
+        SelfResolver::new(self_ty).visit_type_mut(&mut ffi_type);
+
+        let src_to_ffi = visitor.src_to_ffi.expect_or_abort("Defined");
+        let ffi_to_src = visitor.ffi_to_src.expect_or_abort("Defined");
+
+        Self {
+            name,
+            src_type,
+            ffi_type,
+            src_to_ffi,
+            ffi_to_src,
+        }
+    }
+
+    /// Returns true if this argument is an iterator or a slice reference
+    pub fn is_iter_or_slice_ref(&self, mutability: bool) -> bool {
+        match &self.src_type {
+            Type::Reference(type_) => {
+                if !(mutability ^ type_.mutability.is_some()) {
+                    return matches!(*type_.elem, Type::Slice(_));
+                }
+            }
+            Type::ImplTrait(type_) => {
+                assert_eq!(type_.bounds.len(), 1);
+
+                if let syn::TypeParamBound::Trait(trait_) = &type_.bounds[0] {
+                    let trait_name = get_ident(&trait_.path);
+
+                    if !mutability && trait_name == "IntoIterator" {
+                        return true;
+                    }
+
+                    return mutability && trait_name == "ExactSizeIterator";
+                }
+            }
+            _ => return false,
+        }
+
+        false
```

Comment:
Could make result of match itself be the return value.

Reply:
Some ifs don't have an else, so the match statement doesn't always return a value